### ISSUE TITLE
Recommended way to refactor common step definitions?

### ISSUE THREAD

**Issue by ivanschuetz** [cucumber-rs/cucumber#185]
When there are identical steps definitions for different features, what's the best way to refactor them?

From what I've seen so far, each feature/file is connected to a world, which has its own step definitions, so probably it's not possible to reuse directly step definitions (?). Should I just extract the code to regular shared functions?

**Comment by ilslv** [comment]
@ivanschuetz if I understood your question correctly, there is no need to refactor anything in your case. 

Each `Step` function is connected to the particular `World` implementation. When you create `Cucumber` instance linked to a single `World`, all `Feature`s which are matched by path and filter will be matched against `World`. So most of the time you don't need multiple `World`s, as each `World` isn't liked to a single `Feature`.

**Comment by ivanschuetz** [comment]
I see. If I understand you correctly, I'd have to share the same world for these (mostly unrelated) features:

https://github.com/algorand/algorand-sdk-testing/blob/master/features/integration/abi.feature#L5-L6, https://github.com/algorand/algorand-sdk-testing/blob/master/features/integration/applications.feature#L4-L5, 
https://github.com/algorand/algorand-sdk-testing/blob/master/features/integration/assets.feature#L5-L6
(there are more that share these dependencies)

Maintaining all this unrelated state in a single place seems problematic potentially.

**Comment by ilslv** [comment]
@ivanschuetz you can use different `World`s for different `Feature`s by specifying concrete paths in `World::run()` method, but it's not recommended. 

> Maintaining all this unrelated state in a single place seems problematic potentially.

We have 90+ features and 800+ scenarios and it hasn't been a problem for us. To be fare I don't see how it can be problematic, especially when we pass everything by `&mut` ref avoiding unnecessary overhead.

**Comment by ivanschuetz** [comment]
Hmm other of the SDKs do it like that, but I found it difficult to read and work with, as it's difficult to follow what belongs where. Also:

https://cucumber.io/docs/cucumber/state/?sbsearch=World

> It’s important to prevent state created by one scenario from leaking into others. Leaking state makes your scenarios brittle, and difficult to run in isolation.
> 
> ...
>
> Be careful with state
> State can make your steps more tightly coupled and harder to reuse.

Just my "first impression" - just started working with cucumber. Global state like that is often avoided.

That said, I realize that it can be impractical if you have to run the tests using different filters, e.g. specific features vs tags - since with "my approach" you'd have to write a new World for each possible combination.

**Comment by tyranron** [comment]
@ivanschuetz notice, that we don't have global state in this Cucumber implementation particulary. A new fresh `World` is created for each `Scenario` and is used exclusively by it (that's why we have `&mut World` in signatures).

Of course, you still may organize some global state by putting into `World` references to some global statics or other shared state (so the references will leave per scenario, while the state not), but in practice, that's not very useful.

Instead of tending using different `World`s, tend to declare all the `World` necessities close to your `.feature`s. Sharing the same `World` type across different features is more like a feature here, because drastically increases code reusage.

Could you provide a simplified version of the situation you have? Maybe we'll be able to lead you to better modules organization with some concrete tips.

**Comment by ivanschuetz** [comment]
Hey! Yes - putting everything in a single `World` seems to make the most sense.

My initial attempt was to create a separate `World` per feature: https://github.com/manuelmauro/algonaut/blob/main/tests/features_runner.rs
This is the `World` and definitions for one of the features:
https://github.com/manuelmauro/algonaut/blob/main/tests/step_defs/integration/applications.rs
(haven't implement more yet)

But, yeah, as I realized myself above this is pretty inflexible.

These are the features:
https://github.com/algorand/algorand-sdk-testing/tree/master/features/integration
Basically most share the background (client, wallet, related setup)

Not sure a simplified version is needed here - it might be simple enough at high level?

**Comment by ilslv** [comment]
@ivanschuetz it looks like your question has been resolved, can we close this now?
